File: app/lib/features/chaptarr/logic/book_identity.dart
import '../data/chaptarr_models.dart';
import '../../request/data/book_ownership.dart';

final _digits = RegExp(r'^[0-9]+$');
final _olWork = RegExp(r'^OL[0-9]+W$');
final _olEdition = RegExp(r'^OL[0-9]+M$');

String? _providerKey(String kind, String prefix, String? value, RegExp shape) {
  var id = value?.trim() ?? '';
  if (id.contains(':')) {
    if (id.split(':').first.toLowerCase() != prefix) return null;
    id = id.substring(id.indexOf(':') + 1);
  }
  id = id.toUpperCase();
  if (!shape.hasMatch(id) || id == '0') return null;
  if (shape == _digits) id = id.replaceFirst(RegExp(r'^0+'), '');
  return id.isEmpty ? null : '$kind:$id';
}

Set<String> nativeBookIdentityKeys(String? foreignId) {
  final id = foreignId?.trim() ?? '';
  if (id.isEmpty) return {};
  final prefix = id.split(':').first.toLowerCase();
  final typed = switch (prefix) {
    'gr' => _providerKey('gr-work', 'gr', id, _digits),
    'hc' => _providerKey('hc-book', 'hc', id, _digits),
    'ol' => _providerKey('ol-work', 'ol', id, _olWork),
    _ => null,
  };
  return {'native-book:$id', if (typed != null) typed};
}

/// Same typed keys as the server's chaptarr.Book.IdentityKeys. Goodreads works
/// and editions have separate namespaces even when their numeric IDs agree.
Set<String> bookIdentityKeys(ChaptarrBook book) {
  final keys = nativeBookIdentityKeys(book.foreignBookId);
  void add(String kind, String prefix, String? id, RegExp shape) {
    final key = _providerKey(kind, prefix, id, shape);
    if (key != null) keys.add(key);
  }

  add('gr-work', 'gr', book.goodreadsWorkId, _digits);
  add('gr-edition', 'gr', book.goodreadsBookId, _digits);
  add('hc-book', 'hc', book.hardcoverBookId, _digits);
  add('ol-work', 'ol', book.openLibraryWorkId, _olWork);
  for (final edition in [
    ChaptarrEdition(id: 0, foreignEditionId: book.foreignEditionId),
    ...book.editions
  ]) {
    add('gr-edition', 'gr', edition.goodreadsEditionId, _digits);
    add('ol-edition', 'ol', edition.openLibraryEditionId, _olEdition);
    add(
        'hc-edition',
        'hc',
        edition.hardcoverEditionId?.replaceFirst('hc:edition:', 'hc:'),
        _digits);
    final foreign = edition.foreignEditionId ?? '';
    if (foreign.toLowerCase().startsWith('gr:')) {
      add('gr-edition', 'gr', foreign, _digits);
    }
    if (foreign.toLowerCase().startsWith('ol:')) {
      add('ol-edition', 'ol', foreign, _olEdition);
    }
    if (foreign.toLowerCase().startsWith('hc:edition:')) {
      add('hc-edition', 'hc', foreign.substring('hc:edition:'.length), _digits);
    }
    for (final isbn in [edition.isbn13, edition.isbn10]) {
      final valid = normalizedBookISBN(isbn);
      if (valid != null) keys.add('isbn:$valid');
    }
  }
  return keys;
}

String? normalizedBookISBN(String? value) {
  var isbn = (value ?? '').toUpperCase().replaceAll(RegExp(r'[-\s]'), '');
  if (isbn.length == 10) {
    var sum = 0;
    for (var i = 0; i < 10; i++) {
      final n = i == 9 && isbn[i] == 'X' ? 10 : int.tryParse(isbn[i]);
      if (n == null) return null;
      sum += (10 - i) * n;
    }
    if (sum % 11 != 0) return null;
    isbn = '978${isbn.substring(0, 9)}';
    sum = 0;
    for (var i = 0; i < 12; i++) {
      sum += int.parse(isbn[i]) * (1 + 2 * (i % 2));
    }
    return '$isbn${(10 - sum % 10) % 10}';
  }
  if (isbn.length != 13 ||
      !_digits.hasMatch(isbn) ||
      !(isbn.startsWith('978') || isbn.startsWith('979'))) {
    return null;
  }
  var sum = 0;
  for (var i = 0; i < 13; i++) {
    sum += int.parse(isbn[i]) * (1 + 2 * (i % 2));
  }
  return sum % 10 == 0 ? isbn : null;
}

class BookLibraryMatch {
  final OwnedTitle? title;
  final bool ambiguous;
  const BookLibraryMatch({this.title, this.ambiguous = false});
}

/// Exact native identity wins. Otherwise all shared explicit identifiers must
/// name a single library title. Neither title nor author text is evidence.
BookLibraryMatch matchBookToLibrary(
    ChaptarrBook book, List<OwnedTitle> library) {
  final foreignId = book.foreignBookId?.trim() ?? '';
  for (final title in library) {
    if (foreignId.isNotEmpty && title.foreignBookId == foreignId) {
      return BookLibraryMatch(title: title);
    }
  }
  final keys = bookIdentityKeys(book);
  final matches = <String, OwnedTitle>{};
  for (final title in library) {
    if (title.foreignBookId.isEmpty) continue;
    final candidates = {
      ...title.identityKeys,
      ...nativeBookIdentityKeys(title.foreignBookId)
    };
    if (keys.any(candidates.contains)) matches[title.foreignBookId] = title;
  }
  if (matches.length > 1) return const BookLibraryMatch(ambiguous: true);
  return BookLibraryMatch(title: matches.values.firstOrNull);
}

File: server/internal/chaptarr/book_identity.go
package chaptarr

import (
	"bytes"
	"encoding/json"
	"errors"
	"regexp"
	"sort"
	"strings"
)

// ProviderID accepts Chaptarr's string and numeric provider IDs without
// converting large integers through float64. Malformed metadata is not identity.
type ProviderID string

func (id *ProviderID) UnmarshalJSON(data []byte) error {
	*id = ""
	if bytes.Equal(data, []byte("null")) {
		return nil
	}
	var text string
	if json.Unmarshal(data, &text) == nil {
		*id = ProviderID(strings.TrimSpace(text))
	} else if digitsID.Match(data) {
		*id = ProviderID(string(data))
	}
	return nil
}

var digitsID = regexp.MustCompile(`^[0-9]+$`)
var olWorkID = regexp.MustCompile(`^OL[0-9]+W$`)
var olEditionID = regexp.MustCompile(`^OL[0-9]+M$`)

// IdentityKeys use separate namespaces for works and editions. In particular,
// gr:123 as a native book/work is NOT Goodreads edition 123. Native Chaptarr's
// BookResource.BuildForeignBookId uses hc/gr/ol work IDs (not facade IDs).
func (b Book) IdentityKeys() []string {
	keys := NativeBookIdentityKeys(b.ForeignBookID)
	add := func(kind, prefix string, id ProviderID, shape *regexp.Regexp) {
		if key := providerKey(kind, prefix, string(id), shape); key != "" {
			keys = append(keys, key)
		}
	}
	add("gr-work", "gr", b.GoodreadsWorkID, digitsID)
	add("gr-edition", "gr", b.GoodreadsBookID, digitsID)
	add("hc-book", "hc", b.HardcoverBookID, digitsID)
	add("ol-work", "ol", b.OpenLibraryWorkID, olWorkID)
	editions := append([]Edition{{ForeignEditionID: b.ForeignEditionID}}, b.Editions...)
	for _, e := range editions {
		add("gr-edition", "gr", e.GoodreadsEditionID, digitsID)
		add("ol-edition", "ol", e.OpenLibraryEditionID, olEditionID)
		hcEdition := strings.Replace(string(e.HardcoverEditionID), "hc:edition:", "hc:", 1)
		add("hc-edition", "hc", ProviderID(hcEdition), digitsID)
		// Only typed, validated foreign editions contribute an alias.
		if strings.HasPrefix(strings.ToLower(e.ForeignEditionID), "gr:") {
			add("gr-edition", "gr", ProviderID(e.ForeignEditionID), digitsID)
		}
		if strings.HasPrefix(strings.ToLower(e.ForeignEditionID), "ol:") {
			add("ol-edition", "ol", ProviderID(e.ForeignEditionID), olEditionID)
		}
		if strings.HasPrefix(strings.ToLower(e.ForeignEditionID), "hc:edition:") {
			add("hc-edition", "hc", ProviderID(e.ForeignEditionID[len("hc:edition:"):]), digitsID)
		}
		for _, isbn := range []string{e.ISBN13, e.ISBN10} {
			if valid := NormalizeISBN(isbn); valid != "" {
				keys = append(keys, "isbn:"+valid)
			}
		}
	}
	return UniqueIdentityKeys(keys)
}

func (b LookupResult) IdentityKeys() []string {
	book := Book{ForeignBookID: b.ForeignBookID, ForeignEditionID: b.ForeignEditionID, GoodreadsBookID: b.GoodreadsBookID,
		GoodreadsWorkID: b.GoodreadsWorkID, HardcoverBookID: b.HardcoverBookID, OpenLibraryWorkID: b.OpenLibraryWorkID}
	for _, raw := range b.Editions {
		var edition Edition
		if json.Unmarshal(raw, &edition) == nil {
			book.Editions = append(book.Editions, edition)
		}
	}
	return book.IdentityKeys()
}

func NativeBookIdentityKeys(id string) []string {
	id = strings.TrimSpace(id)
	if id == "" {
		return nil
	}
	keys := []string{"native-book:" + id}
	prefix, _, _ := strings.Cut(strings.ToLower(id), ":")
	var key string
	switch prefix {
	case "gr":
		key = providerKey("gr-work", "gr", id, digitsID)
	case "hc":
		key = providerKey("hc-book", "hc", id, digitsID)
	case "ol":
		key = providerKey("ol-work", "ol", id, olWorkID)
	}
	if key != "" {
		keys = append(keys, key)
	}
	return keys
}

func providerKey(kind, prefix, id string, shape *regexp.Regexp) string {
	id = strings.TrimSpace(id)
	if p, value, has := strings.Cut(id, ":"); has {
		if !strings.EqualFold(p, prefix) {
			return ""
		}
		id = value
	}
	id = strings.ToUpper(id)
	if !shape.MatchString(id) || id == "0" {
		return ""
	}
	if shape == digitsID {
		id = strings.TrimLeft(id, "0")
		if id == "" {
			return ""
		}
	}
	return kind + ":" + id
}

func UniqueIdentityKeys(keys []string) []string {
	set := make(map[string]bool, len(keys))
	for _, key := range keys {
		if key != "" {
			set[key] = true
		}
	}
	out := make([]string, 0, len(set))
	for key := range set {
		out = append(out, key)
	}
	sort.Strings(out)
	return out
}

// NormalizeISBN validates the checksum and returns ISBN-13, so a valid ISBN-10
// and its 978 equivalent identify the same publication. Junk never binds books.
func NormalizeISBN(isbn string) string {
	isbn = strings.ToUpper(strings.Join(strings.Fields(strings.ReplaceAll(isbn, "-", "")), ""))
	if len(isbn) == 10 {
		sum := 0
		for i, c := range isbn {
			n := int(c - '0')
			if i == 9 && c == 'X' {
				n = 10
			} else if n < 0 || n > 9 {
				return ""
			}
			sum += (10 - i) * n
		}
		if sum%11 != 0 {
			return ""
		}
		isbn = "978" + isbn[:9]
		sum = 0
		for i, c := range isbn {
			sum += int(c-'0') * (1 + 2*(i%2))
		}
		return isbn + string(rune('0'+(10-sum%10)%10))
	}
	if len(isbn) != 13 || (!strings.HasPrefix(isbn, "978") && !strings.HasPrefix(isbn, "979")) || !digitsID.MatchString(isbn) {
		return ""
	}
	sum := 0
	for i, c := range isbn {
		sum += int(c-'0') * (1 + 2*(i%2))
	}
	if sum%10 != 0 {
		return ""
	}
	return isbn
}

var ErrBookIdentityAmbiguous = errors.New("book identifiers match more than one library title")

// BookIdentityIndex never merges library records. The value of each key is the
// set of native title groups stating it. A match spanning two groups is unsafe,
// even if one key is unique or the titles happen to agree.
type BookIdentityIndex map[string][]string

func IndexBookIdentities(books []Book) BookIdentityIndex {
	index := BookIdentityIndex{}
	for _, book := range books {
		if book.ForeignBookID == "" {
			continue
		}
		for _, key := range book.IdentityKeys() {
			index[key] = UniqueIdentityKeys(append(index[key], book.ForeignBookID))
		}
	}
	return index
}

func (index BookIdentityIndex) Resolve(foreignID string, keys []string) (string, error) {
	foreignID = strings.TrimSpace(foreignID)
	if len(index["native-book:"+foreignID]) > 0 {
		return foreignID, nil
	}
	matches := map[string]bool{}
	for _, key := range append(NativeBookIdentityKeys(foreignID), keys...) {
		for _, id := range index[key] {
			matches[id] = true
		}
	}
	if len(matches) > 1 {
		return "", ErrBookIdentityAmbiguous
	}
	for id := range matches {
		return id, nil
	}
	return "", nil
}
